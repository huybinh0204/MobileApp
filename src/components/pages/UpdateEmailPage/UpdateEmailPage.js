import { yupResolver } from '@hookform/resolvers/yup';
import { useNavigation, useRoute } from '@react-navigation/native';
import React, { useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { useDispatch, useSelector } from 'react-redux';
import { object, string } from 'yup';
import { Form, FormInput, FormSubmitButton, Toast } from '_components/atoms';
import { SecondaryScreenLayout } from '_components/organisms';
import { EVENTS, EVENT_TYPES, NAVIGATION, REGEX, TOAST_TYPES } from '_constants';
import strings from '_localization';
import { CustomerActions, CustomerSelectors } from '_store/customer';
import { RegisterActions } from '_store/register';
import { FormContainer } from './UpdateEmailPage.styles';

const schema = object().shape({
  email: string().matches(REGEX.email, strings.signUp.invalidEmail).required(),
});

const UpdateEmailPage = () => {
  const { params } = useRoute();
  const { navigate } = useNavigation();
  const dispatch = useDispatch();

  const defaultValues = {
    email: params?.email ?? '',
  };

  const formMethods = useForm({ defaultValues, resolver: yupResolver(schema) });

  const isLoading = useSelector(CustomerSelectors.getIsLoading);
  const customerExternalId = useSelector(CustomerSelectors.getExternalId);
  const updateEmailError = useSelector(CustomerSelectors.getUpdateEmailError);
  const updateEmailSuccess = useSelector(CustomerSelectors.getUpdateEmailSuccess);

  const emailValue = formMethods.watch('email');

  useEffect(() => {
    if (updateEmailSuccess) {
      navigate(NAVIGATION.settings.userInfo);
    }
  }, [updateEmailSuccess, navigate]);

  const handleSubmit = ({ email }) => {
    dispatch(RegisterActions.trackEvent(EVENTS.UPDATE_EMAIL_REQUEST, EVENT_TYPES.TRACK));
    dispatch(CustomerActions.updateEmail(customerExternalId, email));
  };

  const handleCloseToast = () => {
    dispatch(CustomerActions.setUpdateEmailError(null));
    dispatch(CustomerActions.setUpdateEmailSuccess(false));
  };

  return (
    <SecondaryScreenLayout testID="UpdateEmailPage" title={strings.signUp.email.placeholder}>
      <Form formMethods={formMethods}>
        <FormContainer>
          <FormInput
            autoCapitalize="none"
            autoComplete="email"
            keyboardType="email-address"
            name="email"
            placeholder={strings.signUp.email.placeholder}
            testID="newAccountMail"
            textContentType="emailAddress"
          />
          <FormSubmitButton
            disabled={emailValue === ''}
            onSubmit={handleSubmit}
            testID="submit"
            isLoading={isLoading}
          >
            {strings.signUp.continue}
          </FormSubmitButton>
        </FormContainer>
      </Form>
      <Toast
        type={TOAST_TYPES.ERROR}
        header={updateEmailError?.title}
        content={updateEmailError?.description}
        onClose={handleCloseToast}
        show={updateEmailError !== null}
        testID="updateMailToastError"
      />
    </SecondaryScreenLayout>
  );
};

export default UpdateEmailPage;
