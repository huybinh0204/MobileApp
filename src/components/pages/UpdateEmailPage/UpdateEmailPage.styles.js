import styled from '@emotion/native';

export const FormContainer = styled.View(({ theme }) => ({
  flex: 1,
  justifyContent: 'space-between',
  padding: theme.spacing.m,
}));
