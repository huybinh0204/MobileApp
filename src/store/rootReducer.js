import { combineReducers } from 'redux';
import actionWatchReducer from 'redux-action-watch/lib/reducer';
import { resettableReducer } from 'reduxsauce';
import { AccountReducer } from './account';
import { AuthenticationReducer } from './authentication';
import { CustomerReducer } from './customer';
import { FundDirectDepositStoreReducer } from './pageStore/fundDirectDepositStore';
import { RegisterReducer } from './register';
import { TransferReducer } from './transfer';

const resettable = resettableReducer('RESET');

export default combineReducers({
  account: resettable(AccountReducer),
  auth: resettable(AuthenticationReducer),
  customer: resettable(CustomerReducer),
  fundDirectDepositStoreReducer: resettable(FundDirectDepositStoreReducer),
  register: RegisterReducer,
  transfer: resettable(TransferReducer),
  watcher: actionWatchReducer, // Must be last
});
