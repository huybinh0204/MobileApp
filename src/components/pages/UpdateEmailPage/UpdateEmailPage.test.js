import { useNavigation } from '@react-navigation/native';
import { fireEvent, render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { useDispatch } from 'react-redux';
import { createStore } from 'redux';
import strings from '_localization/index';
import rootReducer from '_store/rootReducer';
import ComponentWithProviders from '_test/mocks/ComponentWithProviders';
import { navigationMock } from '_test/mocks/navigation';
import UpdateEmailPage from './UpdateEmailPage';

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: jest.fn(),
}));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

describe('UpdateEmailPage', () => {
  const mockedDispatch = jest.fn();
  const mockedNavigation = navigationMock(jest.fn);
  const mockStore = createStore(rootReducer, { customer: { externalId: 1113 } });

  beforeEach(() => {
    useNavigation.mockReturnValue(mockedNavigation);
    useDispatch.mockReturnValue(mockedDispatch);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should not navigate if the mail input is not correct', async () => {
    const { getByText, getByTestId } = render(
      <ComponentWithProviders component={UpdateEmailPage} store={mockStore} />
    );

    const emailInput = getByTestId('newAccountMail');
    fireEvent.changeText(emailInput, 'testmailcom');

    const sendMail = getByText(strings.signUp.continue);
    fireEvent.press(sendMail);

    await waitFor(() => {
      expect(mockedNavigation.navigate).toBeCalledTimes(0);
    });
  });

  it('should match snapshot', () => {
    const { toJSON } = render(
      <ComponentWithProviders component={UpdateEmailPage} store={mockStore} />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
