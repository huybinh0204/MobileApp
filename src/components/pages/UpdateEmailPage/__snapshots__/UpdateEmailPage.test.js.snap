// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateEmailPage should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(242, 242, 242)",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                  "marginTop": 0,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTSafeAreaView
                        backgroundColor={null}
                        emulateUnlessSupported={true}
                        style={
                          Array [
                            Object {
                              "backgroundColor": "#FAFAFA",
                              "flex": 1,
                            },
                            undefined,
                          ]
                        }
                        testID="UpdateEmailPage"
                      >
                        <View
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "height": 108.5,
                                "justifyContent": "center",
                              },
                              undefined,
                            ]
                          }
                        >
                          <View
                            accessible={true}
                            collapsable={false}
                            focusable={true}
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "left": 36,
                                "opacity": 1,
                                "position": "absolute",
                                "zIndex": 10,
                              }
                            }
                            testID="BackButton"
                          >
                            <SvgMock
                              height={45.5}
                              width={45.5}
                            />
                          </View>
                          <Text
                            allowFontScaling={false}
                            style={
                              Array [
                                Object {
                                  "color": "#1F1D25",
                                  "fontFamily": "Poppins-SemiBold",
                                  "fontSize": 29,
                                  "lineHeight": 40,
                                },
                                undefined,
                              ]
                            }
                          >
                            Email
                          </Text>
                        </View>
                        <View
                          onLayout={[Function]}
                          style={
                            Array [
                              Array [
                                Object {
                                  "flex": 1,
                                },
                                undefined,
                              ],
                              Object {
                                "paddingBottom": 0,
                              },
                            ]
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "flex": 1,
                                  "justifyContent": "space-between",
                                  "padding": 58,
                                },
                                undefined,
                              ]
                            }
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "marginBottom": 87,
                                  },
                                  undefined,
                                ]
                              }
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "center",
                                      "borderRadius": 6,
                                      "borderWidth": 2,
                                      "flexDirection": "row",
                                      "width": "100%",
                                      "zIndex": 0,
                                    },
                                    Object {
                                      "borderColor": undefined,
                                    },
                                  ]
                                }
                              >
                                <BVLinearGradient
                                  colors={
                                    Array [
                                      309817028,
                                      318208316,
                                    ]
                                  }
                                  endPoint={
                                    Object {
                                      "x": 0.75,
                                      "y": 0.5,
                                    }
                                  }
                                  locations={null}
                                  startPoint={
                                    Object {
                                      "x": 0.25,
                                      "y": 0.5,
                                    }
                                  }
                                  style={
                                    Array [
                                      Object {
                                        "borderRadius": 6,
                                        "height": "100%",
                                        "position": "absolute",
                                        "width": "100%",
                                        "zIndex": 1,
                                      },
                                      Object {
                                        "opacity": undefined,
                                      },
                                    ]
                                  }
                                />
                                <Text
                                  allowFontScaling={false}
                                  onLayout={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "bottom": 0,
                                        "fontFamily": "Poppins",
                                        "fontSize": 29,
                                        "left": 25.5,
                                        "lineHeight": 40,
                                        "position": "absolute",
                                        "zIndex": 2,
                                      },
                                      Object {
                                        "color": undefined,
                                        "transform": Array [
                                          Object {
                                            "translateY": undefined,
                                          },
                                          Object {
                                            "translateX": undefined,
                                          },
                                          Object {
                                            "scale": undefined,
                                          },
                                        ],
                                      },
                                    ]
                                  }
                                >
                                  Email
                                </Text>
                                <TextInput
                                  autoCapitalize="none"
                                  autoComplete="email"
                                  keyboardType="email-address"
                                  onBlur={[Function]}
                                  onChangeText={[Function]}
                                  onFocus={[Function]}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#1F1D25",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 29,
                                        "lineHeight": 40,
                                        "paddingBottom": 18,
                                        "paddingHorizontal": 25.5,
                                        "paddingTop": 43.5,
                                        "zIndex": 3,
                                      },
                                      undefined,
                                    ]
                                  }
                                  testID="newAccountMail"
                                  textContentType="emailAddress"
                                  value=""
                                />
                              </View>
                            </View>
                            <View
                              accessibilityState={
                                Object {
                                  "disabled": true,
                                }
                              }
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "alignSelf": "stretch",
                                  "backgroundColor": "#D4D7DB",
                                  "borderRadius": 10,
                                  "opacity": 1,
                                  "paddingHorizontal": 45.5,
                                  "paddingVertical": 31,
                                }
                              }
                              testID="submit"
                            >
                              <Text
                                allowFontScaling={false}
                                disabled={true}
                                style={
                                  Array [
                                    Object {
                                      "color": "#FFFFFF",
                                      "fontFamily": "Poppins-SemiBold",
                                      "fontSize": 25.5,
                                      "lineHeight": 35.5,
                                      "textAlign": "center",
                                    },
                                    undefined,
                                  ]
                                }
                              >
                                Continue
                              </Text>
                            </View>
                          </View>
                          <View
                            onLayout={[Function]}
                            paddingHorizontal={null}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "bottom": 0,
                                  "paddingHorizontal": 58,
                                  "position": "absolute",
                                  "width": "100%",
                                },
                                Object {
                                  "opacity": 1,
                                  "transform": Array [
                                    Object {
                                      "translateY": AnimatedValue {
                                        " __value": 0,
                                      },
                                    },
                                  ],
                                },
                              ]
                            }
                            testID="updateMailToastError"
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "alignSelf": "stretch",
                                    "backgroundColor": "#FFFFFF",
                                    "borderColor": "#F2F3F4",
                                    "borderRadius": 10,
                                    "borderWidth": 0,
                                    "padding": 29,
                                    "shadowColor": "#362E60",
                                    "shadowOffset": Object {
                                      "height": 7,
                                      "width": 0,
                                    },
                                    "shadowOpacity": 0.1,
                                    "shadowRadius": 10,
                                  },
                                  Array [
                                    Object {
                                      "backgroundColor": "#D02B35",
                                      "borderColor": "#D02B35",
                                      "borderWidth": 1,
                                      "width": "100%",
                                    },
                                    undefined,
                                  ],
                                ]
                              }
                              type="error"
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "flex-start",
                                      "flexDirection": "row",
                                      "justifyContent": "space-between",
                                    },
                                    undefined,
                                  ]
                                }
                                testID="toastHeader"
                              >
                                <SvgMock
                                  height={36}
                                  width={36}
                                />
                                <Text
                                  allowFontScaling={false}
                                  numberOfLines={2}
                                  style={
                                    Array [
                                      Object {
                                        "color": "#FFFFFF",
                                        "flex": 1,
                                        "fontFamily": "Poppins-SemiBold",
                                        "fontSize": 21.5,
                                        "lineHeight": 30.5,
                                        "paddingHorizontal": 18,
                                        "paddingTop": 2,
                                      },
                                      undefined,
                                    ]
                                  }
                                  type="error"
                                />
                                <View
                                  accessibilityLabel="Close toast"
                                  accessibilityRole="button"
                                  accessible={true}
                                  collapsable={false}
                                  focusable={true}
                                  onClick={[Function]}
                                  onResponderGrant={[Function]}
                                  onResponderMove={[Function]}
                                  onResponderRelease={[Function]}
                                  onResponderTerminate={[Function]}
                                  onResponderTerminationRequest={[Function]}
                                  onStartShouldSetResponder={[Function]}
                                  style={
                                    Object {
                                      "opacity": 1,
                                    }
                                  }
                                  testID="toastCloseButton"
                                >
                                  <SvgMock
                                    height={20}
                                    style={
                                      Array [
                                        Object {
                                          "height": 65,
                                          "width": 65,
                                        },
                                        undefined,
                                      ]
                                    }
                                    type="error"
                                    width={20}
                                  />
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTSafeAreaView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "height": 44,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 1,
          }
        }
      />
    </View>
  </RNCSafeAreaProvider>
</View>
`;
